texts/0000-applicative-monad-superclass.rst

```diff
@@ -0,0 +1,219 @@
+``MonadPlus``, as already implemented in GHC as of 7.10 (released in early
+2015).
+
+
+
+##########
+Motivation
+##########
+
+
+It's the right thing to do™
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Math. You've all heard this one, it's good and compelling so I don't need to
+spell it out.
+
+
+
+Redundant functions
+~~~~~~~~~~~~~~~~~~~
+
+- ``pure`` and ``return`` do the same thing.
+- ``>>`` and ``*>`` are identical.
+- ``liftM`` and ``liftA`` are ``fmap``. The ``liftM*`` are ``liftA*``, ``<*>``
+  is ``ap``.
+- Prelude's ``sequence`` requres ``Monad`` right now, while ``Applicative`` is
+  sufficient to implement it. The more general version of this issue is
+  captured by ``Data.Traversable``, whose main typeclass implements the *same*
+  functionality twice, namely ``traverse`` and ``mapM``, and ``sequenceA`` and
+  ``sequence``.
+
+That very much violates the “don't repeat yourself” principle, and even more so
+it *forces* the programmer to repetition in order to achieve maximal generality.
+
+
+
+
+Using Functor/Applicative functions in monadic code
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Whenever there's Monad code, you can use ``Functor``/``Applicative`` functions,
+without introducing an additional constraint. Keep in mind that
+“``Functor``/``Applicative`` functions” does not only include what their
+typeclasses define but many more, for example ``void``, ``(<$>)``, ``(<**>)``.
+
+Even if you think you have monadic code, strictly using the least restrictive
+functions may result in something that requires only ``Applicative``. This is
+similar to writing a function that needs ``Int``, but it turns out any
+``Integral`` will do – more polymorphism for free.
+
+
+
+Beginner friendliness
+~~~~~~~~~~~~~~~~~~~~~
+
+How often did you say …
+
+- “A Monad is always an Applicative but due to historical reasons it’s not but
+  you can easily verify it by setting ``pure = return`` and ``(<*>) = ap``”
+- “``liftM`` is ``fmap`` but not really.” – “So when should I use ``fmap`` and
+  when ``liftM``?” - *sigh*
+
+With the new hierarchy, the answer would *always* be “use the least restrictive
+one”.
+
+
+
+
+###############
+Detailed design
+###############
+
+Add the following definitions to the Prelude, and demand their existence in the
+Report:
+
+.. code-block:: haskell
+
+   -- (Unchanged)
+   class  Functor f  where
+
+       -- Minimal complete definition: fmap
+
+       fmap :: (a -> b) -> f a -> f b
+
+       (<$) :: a -> f b -> f a
+       (<$) =  fmap . const
+
+
+
+   -- (Unchanged)
+   class Functor f => Applicative f where
+
+       -- Minimal complete definition: pure and (<*>)
+
+       pure :: a -> f a
+
+       (<*>) :: f (a -> b) -> f a -> f b
+
+       (*>) :: f a -> f b -> f b
+       (*>) x y = const id <$> x <*> y
+
+       (<*) :: f a -> f b -> f a
+       (<*) x y = const <$> x <*> y
+
+
+
+   class Applicative m => Monad m where
+
+       -- Minimal complete definition: (>>=) or join
```

Comment:
Yes, join should be one option for a minimal complete definition, because, math.